x/dualstaking/keeper/delegate_credit.go

```diff
@@ -0,0 +1,107 @@
+func (k Keeper) CalculateCredit(ctx sdk.Context, delegation types.Delegation) (credit sdk.Coin, creditTimestampRet int64) {
+	// Calculate the credit for the delegation
+	currentAmount := delegation.Amount
+	creditAmount := delegation.Credit
+	// handle uninitialized amounts
+	if creditAmount.IsNil() {
+		creditAmount = sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), sdk.ZeroInt())
+	}
+	if currentAmount.IsNil() {
+		// this should never happen, but we handle it just in case
+		currentAmount = sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), sdk.ZeroInt())
+	}
+	currentTimestamp := ctx.BlockTime().UTC()
+	delegationTimestamp := time.Unix(delegation.Timestamp, 0)
+	creditTimestamp := time.Unix(delegation.CreditTimestamp, 0)
+	// we normalize dates before we start the calculation
+	// maximum scope is 30 days, we start with the delegation truncation then the credit
+	monthAgo := currentTimestamp.AddDate(0, 0, -30) // we are doing 30 days not a month a month can be a different amount of days
+	if monthAgo.After(delegationTimestamp) {
+		// in the case the delegation wasn't changed for 30 days or more we truncate the timestamp to 30 days ago
+		// and disable the credit for older dates since they are irrelevant
+		delegationTimestamp = monthAgo
+		creditTimestamp = delegationTimestamp
+		creditAmount = sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), sdk.ZeroInt())
+	} else if monthAgo.After(creditTimestamp) {
+		// delegation is less than 30 days, but credit might be older, so truncate it to 30 days
+		creditTimestamp = monthAgo
+	}
+
+	creditDelta := int64(0) // hours
+	if delegation.CreditTimestamp == 0 || creditAmount.IsZero() {
+		// in case credit was never set, we set it to the delegation timestamp
+		creditTimestamp = delegationTimestamp
+	} else if creditTimestamp.Before(delegationTimestamp) {
+		// calculate the credit delta in hours
+		creditDelta = (delegationTimestamp.Unix() - creditTimestamp.Unix()) / hourSeconds
+	}
+
+	amountDelta := int64(0) // hours
+	if !currentAmount.IsZero() && delegationTimestamp.Before(currentTimestamp) {
+		amountDelta = (currentTimestamp.Unix() - delegationTimestamp.Unix()) / hourSeconds
+	}
+
+	// creditDelta is the weight of the history and amountDelta is the weight of the current amount
+	// we need to average them and store it in the credit
+	totalDelta := creditDelta + amountDelta
+	if totalDelta == 0 {
+		return sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), sdk.ZeroInt()), currentTimestamp.Unix()
+	}
+	credit = sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), currentAmount.Amount.MulRaw(amountDelta).Add(creditAmount.Amount.MulRaw(creditDelta)).QuoRaw(totalDelta))
+	return credit, creditTimestamp.Unix()
+}
+
+// this function takes the delegation and returns it's credit within the last 30 days
+func (k Keeper) CalculateMonthlyCredit(ctx sdk.Context, delegation types.Delegation) (credit sdk.Coin) {
+	credit, creditTimeEpoch := k.CalculateCredit(ctx, delegation)
```

Comment:
why is it called "creditTimeEpoch"?

Reply:
because it's the credit timestamp in Epoch format